processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -107,104 +105,45 @@ public int compare(String o1, String o2)
+  {
+    byte versionFromBuffer = buffer.get();
 
-        final int startOffset;
-        final int endOffset;
+    if (VERSION_ONE == versionFromBuffer) {
+      return createGenericIndexedVersionOne(buffer, strategy);
+    } else if (VERSION_TWO == versionFromBuffer) {
+      return createGenericIndexedVersionTwo(buffer, strategy, fileMapper);
+    }
 
-        if (index == 0) {
-          startOffset = 4;
-          endOffset = headerBuffer.getInt(0);
-        } else {
-          int headerPosition = (index - 1) * Ints.BYTES;
-          startOffset = headerBuffer.getInt(headerPosition) + Ints.BYTES;
-          endOffset = headerBuffer.getInt(headerPosition + Ints.BYTES);
-        }
-        return _get(valueBuffer.asReadOnlyBuffer(), startOffset, endOffset);
-      }
-    };
+    throw new IAE("Unknown version [%s]", versionFromBuffer);
   }
 
-  // used for multiple file version, v2.
-  GenericIndexed(
-      List<ByteBuffer> valueBuffs,
-      ByteBuffer headerBuff,
-      ObjectStrategy<T> strategy,
-      boolean allowReverseLookup,
-      int logBaseTwoOfElementsPerValueFile,
-      int numWritten
-  )
+  public static <T> GenericIndexed<T> fromArray(T[] objects, ObjectStrategy<T> strategy)
   {
-    this.strategy = strategy;
-    this.allowReverseLookup = allowReverseLookup;
-    this.valueBuffers = valueBuffs;
-    this.headerBuffer = headerBuff;
-    this.size = numWritten;
-    this.logBaseTwoOfElementsPerValueFile = logBaseTwoOfElementsPerValueFile;
-    headerBuffer.order(ByteOrder.nativeOrder());
-    bufferIndexed = new BufferIndexed()
-    {
-      @Override
-      public T get(int index)
-      {
-        int fileNum = index >> GenericIndexed.this.logBaseTwoOfElementsPerValueFile;
-        final ByteBuffer copyBuffer = valueBuffers.get(fileNum).asReadOnlyBuffer();
-
-        checkIndex(index, size);
+    return fromIterable(Arrays.asList(objects), strategy);
+  }
 
-        final int startOffset;
-        final int endOffset;
-        int relativePositionOfIndex = index & ((1 << GenericIndexed.this.logBaseTwoOfElementsPerValueFile) - 1);
-        if (relativePositionOfIndex == 0) {
-          int headerPosition = index * Ints.BYTES;
-          startOffset = 4;
-          endOffset = headerBuffer.getInt(headerPosition);
-        } else {
-          int headerPosition = (index - 1) * Ints.BYTES;
-          startOffset = headerBuffer.getInt(headerPosition) + 4;
-          endOffset = headerBuffer.getInt(headerPosition + 4);
-        }
-        return _get(copyBuffer, startOffset, endOffset);
-      }
-    };
+  public static <T> GenericIndexed<T> fromIterable(Iterable<T> objectsIterable, ObjectStrategy<T> strategy)
+  {
+    return fromIterableVersionOne(objectsIterable, strategy);
   }
 
-  public static int getNumberOfFilesRequired(int bagSize, long numWritten)
+  static int getNumberOfFilesRequired(int bagSize, long numWritten)
```

Comment:
`getNumberOfFilesRequired()` makes sense only in `GenericIndexed` and `GenericIndexedWriter`. It would be private if `GenericIndexedWriter` would be a nested class of `GenericIndexed`